controllers/workspace_controller_variables.go

```diff
@@ -5,185 +5,162 @@ package controllers
 
 import (
 	"context"
+	"crypto/sha256"
+	"encoding/hex"
 	"fmt"
 
+	tfc "github.com/hashicorp/go-tfe"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 
-	"github.com/google/go-cmp/cmp"
-	"github.com/google/go-cmp/cmp/cmpopts"
-	tfc "github.com/hashicorp/go-tfe"
 	appv1alpha2 "github.com/hashicorp/terraform-cloud-operator/api/v1alpha2"
 )
 
-// HELPERS
-
-// getVariablesToCreate returns a map of variables that need to be created, i.e. exist in manifest, but absent in workspace.
-func getVariablesToCreate(instanceVariables, workspaceVariables map[string]tfc.Variable) map[string]tfc.Variable {
-	return varDifference(instanceVariables, workspaceVariables)
-}
-
-// getVariablesToDelete returns a map of variables that need to be deleted, i.e. exist in workspace, but absent in manifest.
-func getVariablesToDelete(instanceVariables, workspaceVariables map[string]tfc.Variable) map[string]tfc.Variable {
-	return varDifference(workspaceVariables, instanceVariables)
-}
-
-// getVariablesToUpdate returns a map of variables that exist in both
-// the manifest and workspace but have differing values.
-//
-// NOTE We cannot read the value of sensitive variables from the
-// Terraform Cloud API, so sensitive variables will always need to be updated.
-func getVariablesToUpdate(instanceVariables, workspaceVariables map[string]tfc.Variable) map[string]tfc.Variable {
-	variables := make(map[string]tfc.Variable)
-
-	if len(instanceVariables) == 0 || len(workspaceVariables) == 0 {
-		return variables
-	}
-
-	for ik, iv := range instanceVariables {
-		if wv, ok := workspaceVariables[ik]; ok {
-			if !isNoLongerSensitive(iv.Sensitive, wv.Sensitive) && !cmp.Equal(iv, wv, cmpopts.IgnoreFields(tfc.Variable{}, "ID", "VersionID", "Workspace")) {
-				iv.ID = wv.ID
-				variables[iv.Key] = iv
-			}
-		}
+// variableValueID calculates a hash of a variable.
+func variableValueID(v tfc.Variable) string {
+	hash := sha256.New()
+	hv := appv1alpha2.Variable{
+		Name:        v.Key,
+		Description: v.Description,
+		HCL:         v.HCL,
+		Sensitive:   v.Sensitive,
+		Value:       v.Value,
 	}
+	hash.Write([]byte(fmt.Sprintf("%+v", hv)))
```

Comment:
Nothing wrong with this approach, but I think `fmt.Sprintf` doesn't offer any output stability guarantees.
One simple way serialise the struct to a guaranteed stable output is https://pkg.go.dev/encoding/gob

To use it, you would do this:
```suggestion
	gob.NewEncoder(hash).Encode(hv)
```

No big deal if we keep Sprintf, but just wanted to mention this option.

Reply:
Thanks, I didn't know about `gob`. Switched to it.